### PR TITLE
Upgrade openai package

### PR DESCRIPTION
**Pull request by chisingh**
Upgrade to openai v1 python package

### CODE REVIEW

**Comment by matheper** [comment]
The `total_cost` has changed because DLN relies on the number of tokens. `text-davinci-003` and `gpt-3.5-turbo-instruct` use different tokenizers/encoders.

```
In [1]: import tiktoken

In [2]: gpt_3 = tiktoken.encoding_for_model("text-davinci-003")

In [3]: gpt_3_5 = tiktoken.encoding_for_model("gpt-3.5-turbo-instruct")

In [4]: gpt_3.encode("What is the second-largest city in Canada?")
Out[4]: [2061, 318, 262, 1218, 12, 28209, 1748, 287, 3340, 30]

In [5]: gpt_3_5.encode("What is the second-largest city in Canada?")
Out[5]: [3923, 374, 279, 2132, 68067, 3363, 304, 7008, 30]

In [6]: gpt_3.name
Out[6]: 'p50k_base'

In [7]: gpt_3_5.name
Out[7]: 'cl100k_base'
```

File: dln/operator.py [hunk 1]
```diff
@@ -16,21 +18,20 @@
...
         stop=stop_after_attempt(max_attempts),
         wait=wait_exponential(multiplier=1, min=min_wait, max=max_wait),
         retry=(
-            retry_if_exception_type(openai.error.Timeout)
-            | retry_if_exception_type(openai.error.APIError)
-            | retry_if_exception_type(openai.error.APIConnectionError)
-            | retry_if_exception_type(openai.error.RateLimitError)
-            | retry_if_exception_type(openai.error.ServiceUnavailableError)
+            retry_if_exception_type(openai.Timeout)
+            | retry_if_exception_type(openai.APIError)
```
Comment (MarcCote) [inline dln/operator.py:34]: Why discard the `APIConnectionError` and `RateLimitError`?

File: dln/operator.py [hunk 2]
```diff
@@ -167,8 +168,9 @@ def __init__(self, model_name: str = "text-davinci-003", **generation_options):
         super().__init__(model_name, **generation_options)
         engine_for_encoder = self.engine.replace("gpt-35", "gpt-3.5")
         self.encoder = instantiate_tokenizer(engine_for_encoder)
-        openai.api_version = os.environ.get('OPENAI_API_VERSION')
         self._has_logprobs = self.engine in self.LOGPROBS_MODELS
+        self.client = openai.OpenAI()
+        self.aclient = openai.AsyncOpenAI()
```
Comment (MarcCote) [inline dln/operator.py:173]: We probably needs to deal with Azure here.

e.g. I used this
`client = openai.AzureOpenAI() if openai.api_type == "azure" else openai.OpenAI()`

File: dln/operator.py [hunk 3]
```diff
@@ -233,25 +233,25 @@ def _get_completion_response(
...
-                    )
-            except openai.InvalidRequestError as err:
-                self._log_invalid_request_error_message(err, prompt)
-            raise e
+                    _ = self.client.completions.create(model=self.engine,
+                    prompt=prompt,
+                    logprobs=top_logprobs or 1,
+                    **kwargs)
+            except openai.APIError as err:
+                if err.type == 'invalid_request_error':
```
Comment (MarcCote) [inline dln/operator.py:249]: Is that documented somewhere? I think I couldn't find it before.

File: dln/operator.py [hunk 4]
```diff
@@ -324,15 +324,16 @@ class VLLM(LLM):
     def __init__(self, model_name: str, **generation_options):
         super().__init__(model_name, **generation_options)
         self.encoder = instantiate_tokenizer(model_name)
+        self.aclient = openai.AsyncOpenAI()
```
Comment (MarcCote) [inline dln/operator.py:327]: Same as above. Needs to deal with Azure version.

File: requirements.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 packaging==23.1
 torch==2.0.0
-openai==0.27.0
+openai==1.14.1
```
Comment (MarcCote) [inline requirements.txt:3]: All my comments are based on the knowledge of 1.0.1, so please let me know if things have change, e.g. dealing with Azure or the err_type.

**Review by MarcCote — CHANGES_REQUESTED** [review]
Overall looks good. I left a few comments/questions.

File: dln/operator.py (hunk 1 above)
Comment (chisingh) [reply dln/operator.py:34]: Good catch, reverted.

File: dln/operator.py (hunk 3 above)
Comment (chisingh) [reply dln/operator.py:251]: I have verified that it is indeed the error type receivd from api after sending an invalid request.

**Comment by chisingh** [comment]
> Overall looks good. I left a few comments/questions.

Thank you.

**Review by MarcCote — APPROVED** [review]

**Comment by MarcCote** [comment]
Looks great. I let @matheper merge it.

**Comment by matheper** [comment]
Can you check how to instantiate multiple models within the same execution, receiving the connection details when you instantiate/registry a model.

They used to be provided in the kwargs of completion/chat_completion, but now they go to the OpenAI object instantiation.

For example:

```
phi_2 = llm_registry.register(
    "microsoft/phi-2",
    "microsoft/phi-2",
    api_base="http://something_here",
    api_key="key",
)
gpt = GPT(
    "gpt-4",
    "gpt-4",
    api_base="http://something_else",
    api_key="another_key",
)
```

**Review by matheper — APPROVED** [review]

**Pull request merged**